Building a Quarkus application with Java 12+ can result in failure to start it in Java 8 runtime

Quarkus applications are generated to be runnable against a minimum Java 8 version. However, there appears to be an issue where if a Quarkus application is built using Java 8+ (especially Java 12 or higher) and then that application is started on Java 8, then it runs into an exception. One such exception is:

```
Exception in thread "main" java.lang.ExceptionInInitializerError
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:27)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl1.<clinit>(ApplicationImpl1.zig:325)
	... 1 more
Caused by: java.lang.NoClassDefFoundError: java/lang/constant/ConstantDesc
	at io.quarkus.arc.runtime.QuarkusConfigProducer_ProducerMethod_produceStringConfigProperty_75511873be6a3e578c4555b07b03c7231f99c7bc_Bean.<init>(QuarkusConfigProducer_ProducerMethod_produceStringConfigProperty_75511873be6a3e578c4555b07b03c7231f99c7bc_Bean.zig:257)
	at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Default_ComponentsProvider.zig:81)
	at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:102)
	at io.quarkus.arc.Arc.initialize(Arc.java:19)
	at io.quarkus.arc.runtime.ArcRecorder.getContainer(ArcRecorder.java:34)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources15.deploy_0(ArcProcessor$generateResources15.zig:72)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources15.deploy(ArcProcessor$generateResources15.zig:36)
	at io.quarkus.runner.ApplicationImpl1.<clinit>(ApplicationImpl1.zig:295)
	... 1 more
Caused by: java.lang.ClassNotFoundException: java.lang.constant.ConstantDesc
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 9 more
```
It looks like we are missing the use of `-release` option that should be configured in our Maven (and Gradle?) compiler plugins to prevent `javac` from using Java 8+ APIs in the generated class files.

I get this as well -- having Java13, but using maven-compiler-plugin::release==1, and then running it inside Docker with "FROM fabric8/java-alpine-openjdk11-jre"

Exception in thread "main" java.lang.ExceptionInInitializerError
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:27)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:455)
	... 1 more
Caused by: java.lang.RuntimeException: Failed to initialize Arc
	at io.quarkus.arc.Arc.initialize(Arc.java:26)
	at io.quarkus.arc.runtime.ArcRecorder.getContainer(ArcRecorder.java:34)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources22.deploy_0(ArcProcessor$generateResources22.zig:72)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources22.deploy(ArcProcessor$generateResources22.zig:36)
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:417)
	... 1 more
Caused by: java.lang.ClassNotFoundException: java.lang.constant.ConstantDesc
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at io.smallrye.config.inject.ConfigProducer_ProducerMethod_produceStringConfigProperty_75511873be6a3e578c4555b07b03c7231f99c7bc_Bean.<init>(ConfigProducer_ProducerMethod_produceStringConfigProperty_75511873be6a3e578c4555b07b03c7231f99c7bc_Bean.zig:308)
	at io.quarkus.arc.setup.Default_ComponentsProvider.addBeans1(Default_ComponentsProvider.zig:386)
	at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Default_ComponentsProvider.zig:38)
	at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:102)
	at io.quarkus.arc.Arc.initialize(Arc.java:20)
	... 5 more

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Types.java
@@ -10,6 +10,7 @@ import io.quarkus.arc.impl.WildcardTypeImpl;
 import io.quarkus.gizmo.BytecodeCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
@@ -44,6 +45,14 @@ final class Types {
     static final Logger LOGGER = Logger.getLogger(Types.class);
 
     private static final Type OBJECT_TYPE = Type.create(DotNames.OBJECT, Kind.CLASS);
+
+    // we ban these interfaces because they are new to Java 12 and are used by java.lang.String which
+    // means that they cannot be included in bytecode if we want to have application built with Java 12+ but targeting Java 8 - 11
+    // actually run on those older versions
+    // TODO:  add a extensible banning mechanism based on predicates if we find that this set needs to grow...
+    private static final Set<DotName> BANNED_INTERFACE_TYPES = new HashSet<>(
+            Arrays.asList(DotName.createSimple("java.lang.constant.ConstantDesc"),
+                    DotName.createSimple("java.lang.constant.Constable")));
 
     private Types() {
     }
@@ -256,6 +265,9 @@ final class Types {
         }
         // Interfaces
         for (Type interfaceType : classInfo.interfaceTypes()) {
+            if (BANNED_INTERFACE_TYPES.contains(interfaceType.name())) {
+                continue;
+            }
             ClassInfo interfaceClassInfo = getClassByName(beanDeployment.getIndex(), interfaceType.name());
             if (interfaceClassInfo != null) {
                 Map<TypeVariable, Type> resolved = Collections.emptyMap();

